File: kotlin/src/main/kotlin/Day01.kt
import java.io.File

object Day01 {
    const val filename = "day01.txt"

    fun partOneSolution(path: String): Int = loadElves(path).max()

    fun partTwoSolution(path: String): Int = loadElves(path).top(3).sum()

    private fun loadElves(path: String): List<Int> =
        parseElves(File(path).readLines())

    private fun parseElves(lines: List<String>): List<Int> =
        groupLines(lines).map { group -> group.sumOf { it.toInt() } }

    private fun groupLines(lines: List<String>): List<List<String>> =
        if (lines.isEmpty()) listOf()
        else {
            val group = lines.takeWhile { it.isNotEmpty() }
            val rest = lines.drop(group.size + 1)
            listOf(group) + groupLines(rest)
        }

    private fun <T : Comparable<T>> List<T>.top(n: Int): List<T> =
        this.sortedDescending().take(n)
}

File: elixir/day01.exs
defmodule Day01 do
  def trimmed_line_stream(path) do
    File.stream!(path)
    |> Stream.map(&String.trim_trailing/1)
  end

  def group_between_value(enum, val) do
    case Enum.split_while(enum, &(&1 != val)) do
      {group, []} -> [group]
      {group, rest} -> [group | group_between_value(Enum.drop(rest, 1), val)]
    end
  end

  def sum_group(group) do
    Enum.map(group, fn s ->
      {n, ""} = Integer.parse(s)
      n
    end)
    |> Enum.sum()
  end

  def load_sums(path) do
    lines = trimmed_line_stream(path)
    groups = group_between_value(lines, "")
    Enum.map(groups, &sum_group/1)
  end

  def part_one_solution(path) do
    load_sums(path) |> Enum.max()
  end

  def part_two_solution(path) do
    load_sums(path) |> Enum.sort(:desc) |> Enum.take(3) |> Enum.sum()
  end
end

IO.inspect(Day01.part_one_solution("../input/day01.txt"))
IO.inspect(Day01.part_two_solution("../input/day01.txt"))
